### PR TITLE
Added CompactionFilterFactory support to RocksJava

### PR DESCRIPTION
**Pull request by adamretter**
This PR also includes some cleanup, bugfixes and refactoring of the Java API. However these are really pre-cursors on the road to CompactionFilterFactory support.

### CODE REVIEW

**Comment by adamretter** [comment]
https://reviews.facebook.net/D61965

**Comment by adamretter** [comment]
@ajkr Any progress on reviewing this please?

**Comment by adamretter** [comment]
@sagar0 I have revived this and fixed it. Could you take a look please, and merge it if you are happy?

File: java/rocksjni/jnicallback.cc [hunk 1]
```diff
@@ -0,0 +1,52 @@
+// Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
+// This source code is licensed under the BSD-style license found in the
+// LICENSE file in the root directory of this source tree. An additional grant
+// of patent rights can be found in the PATENTS file in the same directory.
+//
+// This file implements the callback "bridge" between Java and C++ for
+// rocksdb::Comparator.
```
Comment (sagar0) [inline java/rocksjni/jnicallback.cc]: Copy-paste error in the comment.

File: java/rocksjni/compaction_filter_factory_jnicallback.h [hunk 2]
```diff
@@ -0,0 +1,35 @@
+// Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
+// This source code is licensed under the BSD-style license found in the
+// LICENSE file in the root directory of this source tree. An additional grant
+// of patent rights can be found in the PATENTS file in the same directory.
+//
+// This file implements the callback "bridge" between Java and C++ for
+// rocksdb::Comparator and rocksdb::DirectComparator.
```
Comment (sagar0) [inline java/rocksjni/compaction_filter_factory_jnicallback.h]: Copy-paste error in the comment.

**Comment by adamretter** [comment]
@sagar0 Thanks, I have now made those changes :-)

**Pull request closed without merging**